compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp

```diff
@@ -236,107 +227,332 @@ static bool setPipeliningMarkers(scf::ForOp forOp, bool pipelineStoreStage) {
+  // Ctor.
+  MainLoopInfo(scf::ForOp forOp) : isSchedulable(true) { analyze(forOp); }
+
+  // Iterate through the mainloop and collect `cp.async`, `cp.commit_group`,
+  // `cp.wait_group`, and `barrier` operations. These operations are used to
+  // pipeline the mainloop and cheorograph asyncroncy for a *coarse-grained*
+  // schedule. Additionally, collect the `mma.sync` and `ldmatrix`/`ld.shared`
+  // operations and separate them into kgroups. The information is helpful in
+  // generating an optimal *finer-grained* instruction interleaving of global
+  // memory loads, shared memory loads, and math operations.
+  void analyze(scf::ForOp forOp) {
+    for (Operation& op : forOp.getBody()->getOperations()) {
+      if (op.getNumRegions() > 0) {
+        // Pipeline and schedule the most inner for op ,i.e., the mainloop that
+        // should be a flat region.
+        isSchedulable = false;
+        return;
+      }
+      if (isa<nvgpu::DeviceAsyncCopyOp>(op)) {
+        copyGlobalToSharedOps.insert(&op);
+      }
+      if (isa<nvgpu::DeviceAsyncCreateGroupOp>(op)) {
+        asyncCreateGroupOp.insert(&op);
+      }
+      if (isa<gpu::BarrierOp>(op)) {
+        barrierOps.insert(&op);
+      }
+      if (isa<nvgpu::DeviceAsyncWaitOp>(op)) {
+        asyncWaitOps.insert(&op);
+      }
+      if (isa<nvgpu::MmaSyncOp>(op)) {
+        // MmaSyncOp visitor traverses the chain of mma operations and separates
+        // them into kgroups.
+        vistMmaSyncOp(&op, 0 /*kgroup*/);
+      }
     }
-    if (isa<nvgpu::MmaSyncOp>(op)) {
-      stageCompute.push_back(&op);
+
+    // If one of the ingredients (`cp.async`, `cp.commit_group`,
+    // `cp.wait_group`, `bar.sync`, `mma.sync`, `ldmatrix` or `ld.shared`) for
+    // scheduling is missing, the mainloop cannot be scheduled.
+    if (copyGlobalToSharedOps.empty() || asyncCreateGroupOp.empty() ||
+        asyncWaitOps.empty() || barrierOps.empty() || warpOperations.empty()) {
+      isSchedulable = false;
+      return;
     }
-    if (isa<nvgpu::DeviceAsyncCopyOp, nvgpu::DeviceAsyncCreateGroupOp>(op)) {
-      stageCopyToSharedMemory.push_back(&op);
+
+    // Collect the dependent operations for cp.async in the mainloop order for
+    // coarse-grained software pipeling.
+    for (Operation& op : forOp.getBody()->getOperations()) {
+      if (isa<nvgpu::DeviceAsyncCopyOp>(&op)) {
+        backwardSliceOfDependentOps(copyGlobalToSharedOpDeps, &op,
+                                    forOp.getBody());
+      }
     }
-    if (isa<nvgpu::LdMatrixOp>(op)) {
-      // Prefecth some of the ldmatrix.
-      if (isAOperand(&op)) {
-        numLoadA++;
-        if (numLoadA <= numPrefetchSmemLoadPerOperand) {
-          stagePrefetch.push_back(&op);
-          continue;
+
+    // Collect the dependent operations for `mma.sync` and `ldmatrix/ld.shared`
+    // operations seperated by kgroups for fine-grained instruction scheduling.
+    for (int kgroup = 0; kgroup < getNumberOfKgroups(); ++kgroup) {
```

Comment:
nit: Can you add a comment on why we need to scan for dependency after having done it for async.cp ops. I always lose track of that part.

Reply:
If I switch the order I get the following error that we have also seen when we fixed it. 

```
iree-compile: /mnt/disks/gcloud_workspace/repos/iree/iree_tree_4/iree/third_party/llvm-project/mlir/lib/Dialect/SCF/Transforms/LoopPipelining.cpp:229: auto (anonymous namespace)::LoopPipelinerInternal::analyzeCrossStageValues()::(anonymous class)::operator()(mlir::OpOperand &) const: Assertion `stage > defStage->second' failed.
```

I put in some comment that makes sense to me. Let me know if you have better take or comment here after this debug error information.